compiler/src/dotty/tools/dotc/ast/MainProxies.scala

```diff
@@ -0,0 +1,101 @@
+package dotty.tools.dotc
+package ast
+
+import core._
+import Symbols._, Types._, Contexts._, Decorators._, util.Spans._, Flags._, Constants._
+import StdNames.nme
+import ast.Trees._
+
+/** Generate proxy classes for @main functions.
+ *  A function like
+ *
+ *     @main def f(x: S, ys: T*) = ...
+ *
+ *  would be translated to something like
+ *
+ *     import CommandLineParser._
+ *     class f {
+ *       @static def main(args: Array[String]): Unit =
+ *         try
+ *           f(
+ *             parseArgument[S](args, 0),
+ *             parseRemainingArguments[T](args, 1): _*
+ *           )
+ *         catch case err: ParseError => showError(err)
+ *       }
+ */
+object MainProxies {
+
+  def mainProxies(stats: List[tpd.Tree]) given Context: List[untpd.Tree] = {
+    import tpd._
+    def mainMethods(stats: List[Tree]): List[Symbol] = stats.flatMap {
+      case stat: DefDef if stat.symbol.hasAnnotation(defn.MainAnnot) =>
+        stat.symbol :: Nil
+      case stat @ TypeDef(name, impl: Template) if stat.symbol.is(Module) =>
+        mainMethods(impl.body)
+      case _ =>
+        Nil
+    }
+    mainMethods(stats).flatMap(mainProxy)
+  }
+
+  import untpd._
+  def mainProxy(mainFun: Symbol) given (ctx: Context): List[TypeDef] = {
+    val mainAnnotSpan = mainFun.getAnnotation(defn.MainAnnot).get.tree.span
+
+    val argsRef = Ident(nme.args)
+
+    def addArgs(call: untpd.Tree, formals: List[Type], restpe: Type, idx: Int): untpd.Tree = {
+      val args = formals.zipWithIndex map {
+        (formal, n) =>
+          val (parserSym, formalElem) =
+            if (formal.isRepeatedParam) (defn.CLP_parseRemainingArguments, formal.argTypes.head)
+            else (defn.CLP_parseArgument, formal)
+          val arg = Apply(
+            TypeApply(ref(parserSym.termRef), TypeTree(formalElem) :: Nil),
+            argsRef :: Literal(Constant(idx + n)) :: Nil)
+          if (formal.isRepeatedParam) repeated(arg) else arg
+      }
+      val call1 = Apply(call, args)
+      restpe match {
+        case restpe: MethodType if !restpe.isImplicitMethod =>
+          if (formals.lastOption.getOrElse(NoType).isRepeatedParam)
+            ctx.error(s"varargs parameter of @main method must come last", mainFun.sourcePos)
+          addArgs(call1, restpe.paramInfos, restpe.resType, idx + args.length)
+        case _ =>
+          call1
+      }
+    }
+
+    var result: List[TypeDef] = Nil
+    if (!mainFun.owner.isStaticOwner)
+      ctx.error(s"@main method is not statically accessible", mainFun.sourcePos)
+    else {
+      var call = ref(mainFun.termRef)
+      mainFun.info match {
+        case _: ExprType =>
+        case mt: MethodType =>
+          if (!mt.isImplicitMethod) call = addArgs(call, mt.paramInfos, mt.resultType, 0)
```

Comment:
```scala
object Main {
  delegate for Int = 10
  @main def g given (x: Int): Unit = {
    println(s"Hello with $x")
  }
}
```

Fails with:

```scala
-- Error: ../ecosystem/mainmethods/src/main/scala/Main.scala:11:2 --------------
11 |  @main def g given (x: Int): Unit = {
   |  ^
   |no implicit argument of type Int was found for parameter x of method g in object Main
```

Whereas:

```scala
delegate for Int = 10
object Main {
  @main def g given (x: Int): Unit = {
    println(s"Hello with $x")
  }
}
```

Works fine. I suspect this has to do with the fact that `g` is a static method in a class `g` and somehow the inner `delegate` becomes invisible to it. Even in that scenario though, it should be visible IMO.

Reply:
The behavior is actually intended, and follows from the way expansion of @main is specified.
The method `g` is called from a synthetic class `g` which is conceptually in the same compilation unit as `g` but is at the toplevel. So it an see a global import but it cannot see a local given instance.

Reply:
That makes sense, but it seems strange that the compile-time behavior is changed by the annotation.

Why do we allow implicit parameters on `@main` methods? I suppose "it resolves whatever it needs at the definition site" is the only semantics that makes sense, but I don't know another place in Scala where the definition-site implicit scope affects anything. Or can the implicit parameters participate in command-line parsing?

Reply:
The problem is:

```scala
object Main {
  delegate for Int = 10
  @main def g given (x: Int): Unit = {
    println(s"Hello with $x")
  }
}
```

Desugars to:

```scala
object Main {
  delegate for Int = 10
  def g given (x: Int): Unit = {
    println(s"Hello with $x")
  }
}

class g {
  <static> def main(args: Array[String]): Unit = g(...)
}
```

The implicits defined at `Main` are not resolved from `g` because `class g` is ejected from `Main`. We cannot have it nested because `main` method must be in a top-level class.

However:

```scala
object Main {
  delegate for Int = 10
  def g given (x: Int): Unit = {
    println(s"Hello with $x")
  }
  def g$main$implicitless: Unit = g
}

class g {
  <static> def main(args: Array[String]): Unit = g$main$implicitless
}
```

@odersky do you think something like that will work to mitigate the unintuitive behaviour of the implicits with nested `@main`s?

Reply:
We need to have a boringly simple and obvious mapping from @main function to the generated main class I think that the implemented behavior is not unintuitive at all. The generated main class comes at the end of the compilation unit that contains the @main function. That's all one needs to explain. From that it follows that global imports are visible but local imports are not.

Reply:
In the current implementation, the local imports are not a problem since they will be visible as per the users' intuition:

```scala
object foo {
  val x: Int = 11
}

object Main {
  import foo._
  delegate for Int = 10
  @main def g: Unit = { //given (x: Int): Unit = {
    println(s"Hello with ${the[Int]} and $x")
  }
}
```

Because the imports are used from the _body_ of the @main function, and the function itself does not change its location. The only problem is resolving the implicits – this is because they present the only resolution which happens on the _call_ of the function.  And the call happens in some other location. As far as I know that the only resolution that happens not in the body of the function.

> We need to have a boringly simple and obvious mapping from @main function to the generated main class.

In the current case, we need to teach the users that the behaviour will be unintuitive with respect to implicits (while their scoping intuition will hold for everything else). In the solution I mentioned we will need to tech that a bridge is generated for the main methods to support proper implicits scoping.

In both cases we need to teach them a subtlety about the language. However, if we do generate bridges for implicits, we ensure that the users' intuition holds.

Reply:
I think the generated main function should have only the following body:

```scala
<static> def main(args: Array[String]): Unit = path.to.`main's`.parent.object.main$no$implcits(args)

// In that object
def main$no$implcits(args: Array[String]) = {  // copied from `MainProxies.scala`
          try
            f(
              parseArgument[S](args, 0),
              parseRemainingArguments[T](args, 1): _*
            )
          catch case err: ParseError => showError(err)
}
```

This will ensure that also the nested implicit `FromString` objects are resolved properly.